tests/suites/test_suite_psa_crypto.function

```diff
@@ -3152,6 +3480,1029 @@ exit:
+                                       data_t *nonce,
+                                       data_t *additional_data,
+                                       data_t *input_data,
+                                       int expected_status_arg )
+{
+
+    mbedtls_svc_key_id_t key = MBEDTLS_SVC_KEY_ID_INIT;
+    psa_key_type_t key_type = key_type_arg;
+    psa_algorithm_t alg = alg_arg;
+    psa_aead_operation_t operation;
+    psa_key_attributes_t attributes = PSA_KEY_ATTRIBUTES_INIT;
+    psa_status_t status = PSA_ERROR_GENERIC_ERROR;
+    psa_status_t expected_status = expected_status_arg;
+    unsigned char *output = NULL;
+    unsigned char *ciphertext = NULL;
+    size_t output_size = output_size_arg;
+    size_t ciphertext_size = 0;
+    size_t ciphertext_length = 0;
+    size_t tag_length = 0;
+    uint8_t tag_buffer[PSA_AEAD_TAG_MAX_SIZE];
+
+    PSA_ASSERT( psa_crypto_init( ) );
+
+    psa_set_key_usage_flags( &attributes, PSA_KEY_USAGE_ENCRYPT  );
+    psa_set_key_algorithm( &attributes, alg );
+    psa_set_key_type( &attributes, key_type );
+
+    PSA_ASSERT( psa_import_key( &attributes, key_data->x, key_data->len,
+                                &key ) );
+
+    PSA_ASSERT( psa_get_key_attributes( key, &attributes ) );
+
+    ASSERT_ALLOC( output, output_size );
+
+    ciphertext_size = PSA_AEAD_FINISH_OUTPUT_SIZE( key_type, alg );
+
+    ASSERT_ALLOC( ciphertext, ciphertext_size );
+
+    operation = psa_aead_operation_init( );
+
+    status = psa_aead_encrypt_setup( &operation, key, alg );
+
+    /* If the operation is not supported, just skip and not fail in case the
+     * encryption involves a common limitation of cryptography hardwares and
+     * an alternative implementation. */
+    if( status == PSA_ERROR_NOT_SUPPORTED )
+    {
+        MBEDTLS_TEST_PSA_SKIP_IF_ALT_AES_192( key_type, key_data->len * 8 );
+        MBEDTLS_TEST_PSA_SKIP_IF_ALT_GCM_NOT_12BYTES_NONCE( alg, nonce->len );
+    }
+
+    PSA_ASSERT( status );
+
+    PSA_ASSERT( psa_aead_set_nonce( &operation, nonce->x, nonce->len ) );
+
+    PSA_ASSERT( psa_aead_update_ad( &operation, additional_data->x,
+                                    additional_data->len ) );
+
+    status = psa_aead_update( &operation, input_data->x, input_data->len,
+                              output, output_size, &ciphertext_length );
+
+    TEST_EQUAL( status, expected_status );
+
+    if( expected_status == PSA_SUCCESS )
+    {
+        /* Ensure we can still complete operation. */
+        PSA_ASSERT( psa_aead_finish( &operation, ciphertext, ciphertext_size,
+                                     &ciphertext_length, tag_buffer,
+                                     PSA_AEAD_TAG_MAX_SIZE, &tag_length ) );
+    }
+
+exit:
+    psa_destroy_key( key );
+    mbedtls_free( output );
+    mbedtls_free( ciphertext );
+    psa_aead_abort( &operation );
+    PSA_DONE( );
+}
+/* END_CASE */
+
+/* BEGIN_CASE */
+void aead_multipart_finish_buffer_test( int key_type_arg, data_t *key_data,
```

Comment:
Given that verify is currently unable to output data (verify really only calls finish() ) is this actually worthwhile?

Reply:
You still have the case of an alternative implementation of GCM and thus in the GCM case with `plaintext_size < 15`
psa_aead_verify() returns PSA_ERROR_BUFFER_TOO_SMALL, right ?

Reply:
We currently have no size tests, because psa_aead_verify() doesn't exist, its just a call to finish in the driver wrapper to generate the tag and check it.

So, the only way that extra output could happen is if in the case of GCM we call gcm_finish() at the verify point, which it appears the self test does do. This would indicate that we need to put back psa_aead_verify() (removed over the course of review as it basically didn't do anything) - we are only getting away with what we have because we do not test against GCM_alt. Would you agree?

Reply:
Regardless of how `psa_aead_verify` is implemented, it exists and should be tested.

But we can leave a test gap for cases that Mbed TLS doesn't implement yet: alt implementation or driver that delays output until the finish step, driver with a verify method. Just file issues for identified gaps so that we remember.

Reply:
Currently psa_aead_verify *doesn't* exist, though. I am struggling to see where the extra (buffered) bytes could be written. The test code is just using gcm_finish() to get the tag, I clearly misunderstood that.

Reply:
I do not really understand what you mean by `psa_aead_verify` does not exist. It does: file psa_crypto.c line 3778 and behaves as described by the spec. Maybe you want to say that `mbedtls_psa_aead_verify` does not exist ? Anyway I am ok with Gilles' proposal. Let's do that later.